geomesa-features/geomesa-feature-avro-complex/src/main/scala/org/locationtech/geomesa/features/avro/complex/AvroCodecs.scala

```diff
@@ -0,0 +1,704 @@
+package org.locationtech.geomesa.features.avro.complex
+
+import java.nio.ByteBuffer
+import java.sql.Timestamp
+import java.{net, sql, util}
+import java.util.function.Function
+import java.util.{Calendar, Date, GregorianCalendar, UUID, ArrayList => JArrayList, Collection => JCollection, Collections => JCollections, HashMap => JHashMap}

```

Comment:
Since you mentioned performance, I loaded up the test code in YourKit.  About 10% of the time was spent in ArrayList.addAll.

I'm still trying to understand things.  As an optimization suggestion, I think it'd make sense for the interface to specify the java.util.List interface rather than the implementation.  

Internally, one could use (Scala) collections which would not require array creation/copies.  At the interface with the client, that collection could be wrapped by a JavaConverter or array-copied one time.

Thoughts?